src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCM.java

```diff
@@ -342,9 +345,13 @@ public ListBoxModel doFillGitToolItems() {
         public ListBoxModel doFillServerIdItems(@QueryParameter String serverId) {
             Jenkins.get().checkPermission(Permission.CONFIGURE);
             List<BitbucketServerConfiguration> serverList =
-                    new BitbucketPluginConfiguration().getServerList();
+                    new BitbucketPluginConfiguration().getServerList()
+                            .stream()
+                            .filter(server -> server.getId().equals(serverId) || server.validate().kind == FormValidation.Kind.OK)
```

Comment:
I don't understand this filter nor the check below it. Can you please expand on what this is meant to do?

Reply:
This probably needs explicit commenting to explain, the functionality here is obtuse.
- The filtering of server list by valid kind or matching server ID means that only servers that have been correctly defined in the settings page will appear in this list, stopping users from seeing or selecting invalid ones. I also ensure the id matching the existing configuration is permitted in the edge case the use changes an existing definition to be invalid.
- In addition to the two existing blank conditions, I added a third, in the event that a configuration in use by an SCM is deleted. This case matches an existing bug described in a separate ticket- when that's resolved, this fix should be removed (I'll mark it as such).
The goal of all this is to make sure the server items shown represent as closely as possible the state of the system, and to make sure the user can't select an invalid configuration.

Reply:
The code could be simplified if we merge the two streams
```
            List<Option> options = new BitbucketPluginConfiguration().getServerList()
                    .stream()
                    .filter(server -> server.getId().equals(serverId) || server.validate().kind == FormValidation.Kind.OK)
                    .map(server -> new Option(
                            server.getServerName(),
                            server.getId(),
                            server.getId().equals(serverId)))
                    .collect(Collectors.toList());
            model.addAll(options);
            if (options.isEmpty()) {
                model.includeEmptyValue();
            }
```

Reply:
This snippet means we can't include a blank value if a non-existant server has been selected, but given this will be addressed in a separate ticket, that should be fine.

Reply:
Shouldn't this add it?
```
if (options.isEmpty()) {
                model.includeEmptyValue();
            }
```